src/components/ui/Rotation.tsx

```diff
@@ -17,7 +18,16 @@ export default class Rotation extends Component<RotationProps> {
 
 		return <div className={styles.container}>
 			{events.map((event, index) => {
-				const action = getDataBy(ACTIONS, 'id', event.ability.guid) as TODO
+				let actionId = event.action
+				if (!actionId) {
+					if (event.ability) {
+						actionId = event.ability.guid
+					} else if (event.cause && event.cause.type === 'action') {
+						actionId = event.cause.action
+					}
+				}
```

Comment:
I think this is probably gnarly enough to merit extracting into a separate function, and then using early return or an if/elseif pattern to keep it to a single indent level:
```suggestion
function getActionId(event): number | undefined {
    if (event.action != null) { 
        return event.action
    }
    if (event.ability != null) {
        return event.ability.guid
    }
    if (event.cause != null && event.cause.type === 'action') {
        return event.cause.action
    }
    return undefined
}
```